From 8b7d499bc2eca4ed65871485ebb8acd09e443157 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Fri, 01 Jul 2022 14:24:54 +0300
Subject: [PATCH] [SW-93955] gaudi2: erom: fix erom recovery flow failure

In gaudi2 COMMS protocol is used to read all the firmware
versions and hence the preboot_ver_off_map offset is mapped
as 0. when trying to read this offset(0) is causing the recovery
flow to hung. Therefore read the btl version only in legacy flow.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I837a90f287b1ea8c6eb9aad425ef405065acbcf6
---

diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index c238483..8114b91 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -177,16 +177,18 @@
 void preboot_version(void)
 {
 	int err;
-	u32 btlver = regmap_readl(global_conf.preboot_ver_off_map);
+	u32 btlver;
 	unsigned long offset = (unsigned long) preboot_vers[PREBOOT_BTL_VER];
 	int i = 0, btl_ver_sz = 0;
 	bool is_legacy_boot = hl_is_legacy_boot();
 
 	preboot_puts("\n");
 	if (is_legacy_boot) {
+		btlver = regmap_readl(global_conf.preboot_ver_off_map);
 		preboot_puts_raw(preboot_vers[PREBOOT_BTL_VER]);
 		preboot_puth(btlver, true);
 		i = btl_ver_sz = copy_btl_version();
+		preboot_dev_spec_legacy_vers_setup(PREBOOT_BTL_VER, (u32)(offset & ~SDRAM_BASE));
 	}
 	preboot_puts_raw(preboot_ver_str);
 	preboot_puts("\n");
@@ -198,8 +200,6 @@
 		if (preboot_ver_str[j] == '\0')
 			break;
 	}
-
-	preboot_dev_spec_legacy_vers_setup(PREBOOT_BTL_VER, (u32)(offset & ~SDRAM_BASE));
 
 	err = boot_info_set(HL_BOOT_INFO_PREBOOT_BTL_VERSION, preboot_vers[PREBOOT_BTL_VER],
 			VERSION_MAX_LEN);
